[package] qt/5.15.5: -o qt:qtwebengine=True fails to build on Linux

### Package and Environment Details

* Package Name/Version: **qt/5.15.5**
* Operating System+version: **Fedora 36 x86_64**
* Compiler+version: **GCC 12**
* Conan version: **conan 1.49.0**
* Python version: **Python 3.10.4**


### Conan profile

```ini
[settings]
os=Linux
os_build=Linux
arch=x86_64
arch_build=x86_64
compiler=gcc
compiler.version=12
compiler.libcxx=libstdc++11
build_type=Release
[options]
[conf]
[build_requires]
[env]
```

### Steps to reproduce

There are two problems observed and I believe both are new with the 5.15.5 version. The first is that the build fails with if `-o qt:qtwebengine=True` is passed, but qt:with_dbus is False (the default). This is not detected by the recipe, but causes a build failure:

```
conan install qt/5.15.5@ --build -o qt:qtwebengine=True -o qt:shared=True -o qt:qtwebchannel=True -o qt:qtlocation=True

...
...
( test -e Makefile.gn_run || /home/ts/.conan/data/qt/5.15.5/_/_/build/da47b1d8787266692c9541c14ca069ca109f26b4/build_folder/qtbase/bin/qmake -o Makefile.gn_run /home/ts/.conan/data/qt/5.15.5/_/_/source/qt5/qtwebengine/src/pdf/gn_run.pro ) && make -f Makefile.gn_run 
Project MESSAGE: Running: /home/ts/.conan/data/qt/5.15.5/_/_/build/da47b1d8787266692c9541c14ca069ca109f26b4/build_folder/qtwebengine/src/3rdparty/gn/out/Release/gn gen /home/ts/.conan/data/qt/5.15.5/_/_/build/da47b1d8787266692c9541c14ca069ca109f26b4/build_folder/qtwebengine/src/pdf/release --script-executable=/usr/bin/python2 --args='use_qt=true init_stack_vars=false is_component_build=false is_shared=true enable_debugallocation=false enable_media_remoting=false enable_message_center=false enable_nacl=false enable_remoting=false enable_reporting=false enable_resource_allowlist_generation=false enable_swiftshader=false enable_swiftshader_vulkan=false angle_enable_swiftshader=false enable_web_speech=false enable_widevine=true forbid_non_component_debug_builds=false has_native_accessibility=false safe_browsing_mode=0 skia_use_dawn=false toolkit_views=false treat_warnings_as_errors=false use_allocator_shim=false use_allocator="none" use_custom_libcxx=false chrome_pgo_phase=0 enable_hangout_services_extension=true optimize_webui=false enable_js_type_check=false v8_use_external_startup_data=false strip_absolute_paths_from_debug_symbols=false use_jumbo_build=true jumbo_file_merge_limit=8 jumbo_build_excluded=["browser"] enable_precompiled_headers=true is_official_build=true is_debug=false symbol_level=0 blink_symbol_level=0 remove_v8base_debug_symbols=true use_cups=false use_gio=false use_gnome_keyring=false use_udev=true use_bundled_fontconfig=false use_sysroot=false enable_session_service=false is_cfi=false use_ozone=true use_x11=false ozone_auto_platforms=false ozone_platform_headless=false ozone_platform_external=true ozone_platform="qt" ozone_extra_path="/home/ts/.conan/data/qt/5.15.5/_/_/source/qt5/qtwebengine/src/core/ozone/ozone_extra.gni" use_gold=false use_lld=false is_clang=false custom_toolchain="/home/ts/.conan/data/qt/5.15.5/_/_/build/da47b1d8787266692c9541c14ca069ca109f26b4/build_folder/qtwebengine/src/toolchain:target" host_toolchain="/home/ts/.conan/data/qt/5.15.5/_/_/build/da47b1d8787266692c9541c14ca069ca109f26b4/build_folder/qtwebengine/src/toolchain:host" host_cpu="x64" pkg_config="pkg-config" host_pkg_config="/usr/bin/pkg-config" pdfium_use_system_zlib=true use_system_libpng=true pdfium_use_system_libpng=true use_system_libjpeg=false use_system_freetype=true use_system_harfbuzz=false use_glib=false use_nss_certs=false pdf_enable_v8=false pdf_enable_xfa=false pdf_enable_xfa_bmp=false pdf_enable_xfa_gif=false pdf_enable_xfa_png=false pdf_enable_xfa_tiff=false qtwebengine_target="/home/ts/.conan/data/qt/5.15.5/_/_/build/da47b1d8787266692c9541c14ca069ca109f26b4/build_folder/qtwebengine/src/pdf/release:QtPdf"' --root=/home/ts/.conan/data/qt/5.15.5/_/_/source/qt5/qtwebengine/src/3rdparty/chromium 
ERROR at //build/config/linux/pkg_config.gni:103:17: Script returned non-zero exit code.
    pkgresult = exec_script(pkg_config_script, args, "value")
                ^----------
Current dir: /home/ts/.conan/data/qt/5.15.5/_/_/build/da47b1d8787266692c9541c14ca069ca109f26b4/build_folder/qtwebengine/src/pdf/release/
Command: /usr/bin/python2 /home/ts/.conan/data/qt/5.15.5/_/_/source/qt5/qtwebengine/src/3rdparty/chromium/build/config/linux/pkg-config.py -p pkg-config dbus-1
Returned 1.
stderr:

Package dbus-1 was not found in the pkg-config search path.
Perhaps you should add the directory containing `dbus-1.pc'
to the PKG_CONFIG_PATH environment variable
Package 'dbus-1', required by 'virtual:world', not found
Could not run pkg-config.

See //build/config/linux/dbus/BUILD.gn:12:1: whence it was called.
pkg_config("dbus") {
^-------------------
See //dbus/BUILD.gn:46:22: which caused the file to be included.
  public_configs = [ "//build/config/linux/dbus" ]
                     ^--------------------------
Project ERROR: GN run error!
```

The issue above should ideally be detected by the recipe, in the same way that it checks for e.g. `qt:qtwebchannel=True` and `qt:qtlocation=True`

After passing `-o qt:with_dbus=True` the build gets further, but then stumbles over a font-config issue. Note that `qt:with_fontconfig=True` is default and remains unchanged from the default in the following invocation:

```
conan install qt/5.15.5@ --build -o qt:qtwebengine=True -o qt:shared=True -o qt:qtwebchannel=True -o qt:qtlocation=True -o qt:with_dbus=True

...
...
FAILED: obj/components/services/font/lib/fontconfig_matching.o 
/usr/lib64/ccache/g++ -MMD -MF obj/components/services/font/lib/fontconfig_matching.o.d -DUSE_UDEV -DUSE_AURA=1 -DUSE_NSS_CERTS=1 -DUSE_OZONE=1 -DOFFICIAL_BUILD -DTOOLKIT_QT -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -DNO_UNWIND_TABLES -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -DNDEBUG -DNVALGRIND -DDYNAMIC_ANNOTATIONS_ENABLED=0 -DWEBP_EXTERN=extern -DSK_CODEC_DECODES_PNG -DSK_CODEC_DECODES_WEBP -DSK_ENCODE_PNG -DSK_ENCODE_WEBP -DSK_USER_CONFIG_HEADER=\"../../skia/config/SkUserConfig.h\" -DSK_GL -DSK_CODEC_DECODES_JPEG -DSK_ENCODE_JPEG -DSK_USE_LIBGIFCODEC -DSK_VULKAN_HEADER=\"../../skia/config/SkVulkanConfig.h\" -DSK_VULKAN=1 -DSK_SUPPORT_GPU=1 -DSK_GPU_WORKAROUNDS_HEADER=\"gpu/config/gpu_driver_bug_workaround_autogen.h\" -DU_USING_ICU_NAMESPACE=0 -DU_ENABLE_DYLOAD=0 -DUSE_CHROMIUM_ICU=1 -DU_ENABLE_TRACING=1 -DU_ENABLE_RESOURCE_TRACING=0 -DU_STATIC_IMPLEMENTATION -DICU_UTIL_DATA_IMPL=ICU_UTIL_DATA_FILE -DUCHAR_TYPE=uint16_t -Igen -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/third_party/perfetto/include -Igen/third_party/perfetto/build_config -Igen/third_party/perfetto -Igen -Igen -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/third_party/libwebp/src -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/third_party/skia -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/third_party/libgifcodec -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/third_party/vulkan/include -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/third_party/vulkan_headers/include -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/third_party/abseil-cpp -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/third_party/boringssl/src/include -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/third_party/protobuf/src -Igen/protoc_out -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/third_party/ced/src -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/third_party/icu/source/common -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/third_party/icu/source/i18n -fno-strict-aliasing --param=ssp-buffer-size=4 -fstack-protector -fno-unwind-tables -fno-asynchronous-unwind-tables -fPIC -pipe -pthread -m64 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -Wno-unused-local-typedefs -Wno-maybe-uninitialized -Wno-deprecated-declarations -fno-delete-null-pointer-checks -Wno-comments -Wno-packed-not-aligned -Wno-dangling-else -Wno-missing-field-initializers -Wno-unused-parameter -O2 -fno-ident -fdata-sections -ffunction-sections -fno-omit-frame-pointer -g0 -fvisibility=hidden -std=gnu++14 -Wno-narrowing -Wno-class-memaccess -Wno-attributes -Wno-class-memaccess -Wno-subobject-linkage -Wno-invalid-offsetof -Wno-return-type -Wno-deprecated-copy -fno-exceptions -fno-rtti -fvisibility-inlines-hidden -c ../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/components/services/font/fontconfig_matching.cc -o obj/components/services/font/lib/fontconfig_matching.o
../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/components/services/font/fontconfig_matching.cc:7:10: fatal error: fontconfig/fontconfig.h: No such file or directory
    7 | #include <fontconfig/fontconfig.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

The build failure occurs while building chromium in qtwebengine, so perhaps that the fontconfig library detection/configuration is not passed correctly by the Qt configure script to the chromium part of the build?

Turns out this reproduces with the 5.15.4 package too:

```
$ rm -rf ~/.conan/data/
$ ccache -Cz
$ conan install qt/5.15.4@ --build -o qt:qtwebengine=True -o qt:shared=True -o qt:qtwebchannel=True -o qt:qtlocation=True -o qt:with_dbus=True

...
...
11643/25442] CXX obj/components/services/font/ppapi_fontconfig_matching/ppapi_fontconfig_matching.o
FAILED: obj/components/services/font/ppapi_fontconfig_matching/ppapi_fontconfig_matching.o 
/usr/lib64/ccache/g++ -MMD -MF obj/components/services/font/ppapi_fontconfig_matching/ppapi_fontconfig_matching.o.d -DUSE_UDEV -DUSE_AURA=1 -DUSE_NSS_CERTS=1 -DUSE_OZONE=1 -DOFFICIAL_BUILD -DTOOLKIT_QT -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -DNO_UNWIND_TABLES -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -DNDEBUG -DNVALGRIND -DDYNAMIC_ANNOTATIONS_ENABLED=0 -Igen -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/third_party/perfetto/include -Igen/third_party/perfetto/build_config -Igen/third_party/perfetto -Igen -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/third_party/abseil-cpp -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/third_party/boringssl/src/include -I../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/third_party/protobuf/src -Igen/protoc_out -fno-strict-aliasing --param=ssp-buffer-size=4 -fstack-protector -fno-unwind-tables -fno-asynchronous-unwind-tables -fPIC -pipe -pthread -m64 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -Wno-unused-local-typedefs -Wno-maybe-uninitialized -Wno-deprecated-declarations -fno-delete-null-pointer-checks -Wno-comments -Wno-packed-not-aligned -Wno-dangling-else -Wno-missing-field-initializers -Wno-unused-parameter -O2 -fno-ident -fdata-sections -ffunction-sections -fno-omit-frame-pointer -g0 -fvisibility=hidden -std=gnu++14 -Wno-narrowing -Wno-class-memaccess -Wno-attributes -Wno-class-memaccess -Wno-subobject-linkage -Wno-invalid-offsetof -Wno-return-type -Wno-deprecated-copy -fno-exceptions -fno-rtti -fvisibility-inlines-hidden -c ../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/components/services/font/ppapi_fontconfig_matching.cc -o obj/components/services/font/ppapi_fontconfig_matching/ppapi_fontconfig_matching.o
../../../../../../../source/qt5/qtwebengine/src/3rdparty/chromium/components/services/font/ppapi_fontconfig_matching.cc:8:10: fatal error: fontconfig/fontconfig.h: No such file or directory
    8 | #include <fontconfig/fontconfig.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

It is not clear to me why this is suddenly(?) happening.

--- a/recipes/qt/5.x.x/conandata.yml
+++ b/recipes/qt/5.x.x/conandata.yml
@@ -107,6 +107,8 @@ patches:
       base_path: "qt5/qtwebengine/src/3rdparty"
     - patch_file: "patches/skia-cd397f3.diff"
       base_path: "qt5/qtwebengine/src/3rdparty/chromium/third_party/skia"
+    - patch_file: "patches/0001-Find-fontconfig-using-pkg-config.patch"
+      base_path: "qt5/qtwebengine/src/3rdparty"
   "5.15.5":
     - patch_file: "patches/aa2a39dea5.diff"
       base_path: "qt5/qtbase"
@@ -128,3 +130,5 @@ patches:
       base_path: "qt5/qtwebengine/src/3rdparty"
     - patch_file: "patches/skia-cd397f3.diff"
       base_path: "qt5/qtwebengine/src/3rdparty/chromium/third_party/skia"
+    - patch_file: "patches/0001-Find-fontconfig-using-pkg-config.patch"
+      base_path: "qt5/qtwebengine/src/3rdparty"

--- a/recipes/qt/5.x.x/patches/0001-Find-fontconfig-using-pkg-config.patch
+++ b/recipes/qt/5.x.x/patches/0001-Find-fontconfig-using-pkg-config.patch
@@ -1,0 +1,56 @@
+From fb103459151fcf02706dd5abda95c716a934d610 Mon Sep 17 00:00:00 2001
+From: Allan Sandfeld Jensen <allan.jensen@qt.io>
+Date: Mon, 9 May 2022 10:32:24 +0200
+Subject: [PATCH] Find fontconfig using pkg-config
+
+We already verify that it can be found like that during configure
+
+Task-number: QTBUG-61158
+Change-Id: I569590e96a490c4ed6e6dc560fbd110d86d77956
+Reviewed-by: Michal Klocek <michal.klocek@qt.io>
+(cherry picked from commit 357dcbf7d9510b5282a18e8211e2494f353b4e5e)
+Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
+(cherry picked from commit 43329f9d7f38a5092a9c054c5d450072b236c6f8)
+---
+ chromium/components/services/font/BUILD.gn | 1 +
+ chromium/third_party/fontconfig/BUILD.gn   | 5 +++--
+ 2 files changed, 4 insertions(+), 2 deletions(-)
+
+diff --git chromium/components/services/font/BUILD.gn chromium/components/services/font/BUILD.gn
+index b065131e035..f3e544ed29f 100644
+--- chromium/components/services/font/BUILD.gn
++++ chromium/components/services/font/BUILD.gn
+@@ -42,6 +42,7 @@ if ((is_linux || is_chromeos) && enable_plugins) {
+       "//base:base",
+       "//ppapi/buildflags:buildflags",
+       "//ppapi/c",
++      "//third_party/fontconfig",
+     ]
+   }
+ }
+diff --git chromium/third_party/fontconfig/BUILD.gn chromium/third_party/fontconfig/BUILD.gn
+index 1bee7acb6b9..4cc32d62774 100644
+--- chromium/third_party/fontconfig/BUILD.gn
++++ chromium/third_party/fontconfig/BUILD.gn
+@@ -3,6 +3,7 @@
+ # found in the LICENSE file.
+ 
+ import("//build/config/sanitizers/sanitizers.gni")
++import("//build/config/linux/pkg_config.gni")
+ import("//third_party/fontconfig/fontconfig.gni")
+ 
+ assert(is_linux || is_chromeos)
+@@ -110,8 +111,8 @@ if (use_bundled_fontconfig) {
+     }
+   }
+ } else {
+-  config("fontconfig_config") {
+-    libs = [ "fontconfig" ]
++  pkg_config("fontconfig_config") {
++    packages = [ "fontconfig" ]
+   }
+ 
+   group("fontconfig") {
+-- 
+2.36.1
+

